### PR TITLE
devicestate: read modeenv early and store in devicestate

### PR DESCRIPTION
**Pull request by mvo5**
On UC20 systems the "run-mode" (install,recover,run) is an important
part of the device state. Read it early in the devicemanager code
and keep around. Also avoid registration when in "install" mode.

Extracted parts of #7762 for this.

### CODE REVIEW

File: overlord/devicestate/devicemgr.go [hunk 1]
```diff
@@ -60,6 +61,10 @@ type DeviceManager struct {
 	becomeOperationalBackoff     time.Duration
 	registered                   bool
 	reg                          chan struct{}
+
+	// XXX: is that the best name?
+	// runMode of a UC20 system. "","run","install","recover"
+	runMode string
```
Comment (pedronis) [inline overlord/devicestate/devicemgr.go]: this should probably go at the top of the struct.  About the name runMode is clashing with "run mode" itself, I would call it just mode, or operatingMode

File: overlord/devicestate/devicestate_serial_test.go [hunk 2]
```diff
@@ -1634,3 +1635,44 @@ func (s *deviceMgrSerialSuite) TestDoRequestSerialReregistrationDoubleSerialStre
 	c.Check(chg.Status(), Equals, state.ErrorStatus, Commentf("%s", t.Log()))
 	c.Check(chg.Err(), ErrorMatches, `(?ms).*cannot accept more than a single device serial assertion from the device service.*`)
 }
+
+func (s *deviceMgrSerialSuite) TestDeviceManagerReadsModeenv(c *C) {
```
Comment (pedronis) [inline overlord/devicestate/devicestate_serial_test.go]: this belongs to the devicestate_test.go deviceMgrSuite no ?

File: overlord/devicestate/export_test.go [hunk 3]
```diff
@@ -83,6 +83,14 @@ func SetLastBecomeOperationalAttempt(m *DeviceManager, t time.Time) {
 	m.lastBecomeOperationalAttempt = t
 }
 
+func SetRunMode(m *DeviceManager, mode string) {
+	m.runMode = mode
+}
+
+func RunMode(m *DeviceManager) string {
+	return m.runMode
```
Comment (pedronis) [inline overlord/devicestate/export_test.go]: this is fine for now, at some point we will probably need to have DeviceManager.Mode

File: overlord/devicestate/devicemgr.go (hunk 1 above)
Comment (mvo5) [reply overlord/devicestate/devicemgr.go]: Thanks, I really like operatingMode

**Pull request merged**